sql/core/src/main/scala/org/apache/spark/sql/ForeachWriter.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.annotation.Experimental
+
+/**
+ * :: Experimental ::
+ * A writer to consume data generated by a [[ContinuousQuery]]. Each partition will use a new
+ * deserialized instance, so you usually should do the initialization work in the `open` method.
+ *
+ * @since 2.0.0
+ */
+@Experimental
+trait ForeachWriter[T] extends Serializable {
+
+  /**
+   * Called when starting to process one partition of new data in the executor side. `version` is
+   * for data deduplication. When recovering from a failure, some data may be processed twice. But
+   * it's guarantee that they will be opened with the same "version".
+   *
+   * If this method finds this is a partition from a duplicated data set, it can return `false` to
+   * skip the further data processing. However, `close` still will be called for cleaning up
+   * resources.
+   *
+   * @param version a unique id for data deduplication.
+   * @return a flat that indicates if the data should be processed.
+   */
+  def open(version: Long): Boolean
+
+  /**
+   * Called to process the data in the executor side.
+   */
+  def process(value: T): Unit
+
+  /**
+   * Called when stopping to process one partition of new data in the executor side.
+   *
+   * @param isFailed Whether any error is thrown during processing data.
+   * @param error The error thrown during processing data. if `isFailed` is `true`. Otherwise, it's
+   *              undefined.
+   */
+  def close(isFailed: Boolean, error: Throwable): Unit
```

Comment:
> Its kind of weird that we use Option in the listener but not here? We should decide if thats okay for public APIs or not and be consistent.

It's not `Option` because we don't use `Option` in a Java API.